src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -0,0 +1,152 @@
+            public Mock<ICheckpoints> CheckpointsMock = new Mock<ICheckpoints>();
+            public Mock<IChainState> ChainStateMock = new Mock<IChainState>();
+            public ConsensusSettings ConsensusSettings = new ConsensusSettings();
+
+            public ChainedHeaderTree ChainedHeaderTree;
+
+            public ChainedHeaderTree CreateChainedHeaderTree()
+            {
+                this.ChainedHeaderTree = new ChainedHeaderTree(this.Network, new ExtendedLoggerFactory(), this.ChainedHeaderValidatorMock.Object, this.CheckpointsMock.Object, this.ChainStateMock.Object, this.ConsensusSettings);
+                return this.ChainedHeaderTree;
+            }
+
+            public ChainedHeader ExtendAChain(int count, ChainedHeader chainedHeader = null)
+            {
+                ChainedHeader previousHeader = chainedHeader ?? new ChainedHeader(this.Network.GetGenesis().Header, this.Network.GenesisHash, 0);
+
+                foreach (int index in Enumerable.Range(1, count))
+                {
+                    BlockHeader header  = this.Network.Consensus.ConsensusFactory.CreateBlockHeader();
+                    header.HashPrevBlock = previousHeader.HashBlock;
+                    header.Bits = previousHeader.Header.Bits - 1000; // just increase difficulty.
+                    ChainedHeader newHeader = new ChainedHeader(header, header.GetHash(), previousHeader);
+                    previousHeader = newHeader;
+                }
+
+                return previousHeader;
+            }
+
+            public List<BlockHeader> ChainedHeaderToList(ChainedHeader chainedHeader, int count)
+            {
+                List<BlockHeader> list = new List<BlockHeader>();
+
+                ChainedHeader current = chainedHeader;
+                while (count > 0)
+                {
+                    list.Add(current.Header);
+                    current = current.Previous;
+                    count--;
+                }
+
+                list.Reverse();
+
+                return list;
+            }
+
+            public bool ConnectedHeadersIsEmpty(ConnectedHeaders connectedHeaders)
+            {
+                Assert.NotNull(connectedHeaders);
+
+                return connectedHeaders.Consumed == null 
+                       && connectedHeaders.DownloadTo == null 
+                       && connectedHeaders.DownloadFrom == null;
+            }
+        }
+
+        [Fact]
+        public void ConnectHeaders_HeadersCantConnect_ShouldFail()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree chainedHeaderTree = testContext.CreateChainedHeaderTree();
+
+            Assert.Throws<ConnectHeaderException>(() => chainedHeaderTree.ConnectNewHeaders(1, new List<BlockHeader>(new [] { testContext.Network.GetGenesis().Header})));
+        }
+
+        [Fact]
+        public void ConnectHeaders_NoNewHeadersToConnect_ShouldReturnNothingToDownload()
```

Comment:
Just to note the did this test is not exactly what we wanted in test number 2 Described here https://github.com/stratisproject/StratisBitcoinFullNode/issues/1321#issuecomment-389155904

The difference is that we would like to check that no new headers were actually created in the tree and this test is testing that the return value of the method does not request new blocks to be downloaded.

Reply:
OK yeah so basically we can just add more asserts on the internal dictionary (I'll do it with reflection
)

Reply:
done